orchestrator/lib/coolfhir/contained.go

```diff
@@ -0,0 +1,98 @@
+package coolfhir
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/zorgbijjou/golang-fhir-models/fhir-models/fhir"
+)
+
+var (
+	ErrCareTeamNotFound = errors.New("CareTeam not found in FHIR Bundle")
+	ErrInvalidReference = errors.New("invalid reference, expecting local reference")
+)
+
+type ContainedResource struct {
+	Resource
+	Raw json.RawMessage `json:"-"`
+}
+
+func (r *ContainedResource) UnmarshalJSON(data []byte) error {
+	if err := json.Unmarshal(data, &r.Resource); err != nil {
+		return err
+	}
+
+	r.Raw = data
+
+	return nil
+}
+
+func (r *ContainedResource) MarshalJSON() ([]byte, error) {
+	return r.Raw, nil
+}
+
+func UpdateContainedResource(contained json.RawMessage, id *fhir.Reference, resourceType string, resource any) (json.RawMessage, error) {
+	if !IsLocalReference(id) {
+		return nil, ErrInvalidReference
+	}
+
+	data, err := json.Marshal(resource)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal data: %w", err)
+	}
+
+	var resources []ContainedResource
+
+	if err := json.Unmarshal(contained, &resources); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal contained resources: %w", err)
+	}
+
+	ref := *id.Reference
+
+	for i, containedResource := range resources {
+		if containedResource.ID == ref[1:] && containedResource.Type == resourceType {
+			containedResource.Raw = data
+			resources[i] = containedResource
+
+			return json.Marshal(resources)
+		}
+	}
+
+	return nil, fmt.Errorf("contained resource not found")
+}
+
+func ResolveCareTeam(carePlan *fhir.CarePlan) (*fhir.CareTeam, error) {
```

Comment:
`CareTeamFromCarePlan` or make a more generic function `GetContainedResource` or something, in which you pass the struct to unmarshal into.
Something like we do in `bundle.go`:

`func GetContainedResource(container any, target any, predicate func()) error` (unmarshal into target)